### ISSUE TITLE
FunctionClauseError with 304 Response

### ISSUE THREAD

**Issue by optikfluffel** [appcues/mojito#24]
When playing around with etag and modified_since headers I found something weird.

```elixir
req_headers = [{"If-Modified-Since", "Wed, 22 May 2019 07:43:40 GMT"}]

Mojito.get("https://robynthinks.wordpress.com/feed/", req_headers)
```

results in:

```
{:error, %Mojito.Error{message: nil, reason: :closed}}

[error] GenServer #PID<0.596.0> terminating
** (FunctionClauseError) no function clause matching in Mojito.ConnServer.apply_resp/2
    (mojito) lib/mojito/conn_server.ex:144: Mojito.ConnServer.apply_resp(%{conn: %Mojito.Conn{conn: %Mint.HTTP2{buffer: "", client_settings: %{enable_push: true, max_concurrent_streams: 100, max_frame_size: 16384}, client_settings_queue: {[], []}, decode_table: %Mint.HTTP2.HPACK.Table{entries: [], length: 0, max_table_size: 4096, size: 0}, encode_table: %Mint.HTTP2.HPACK.Table{entries: [{"If-Modified-Since", "Wed, 22 May 2019 07:43:40 GMT"}], length: 1, max_table_size: 4096, size: 78}, headers_being_processed: nil, hostname: "robynthinks.wordpress.com", next_stream_id: 5, open_client_stream_count: 0, open_server_stream_count: 0, ping_queue: {[], []}, port: 443, private: %{}, ref_to_stream_id: %{}, scheme: "https", server_settings: %{enable_push: true, initial_window_size: 65536, max_concurrent_streams: 128, max_frame_size: 16777215, max_header_list_size: :infinity}, socket: {:sslsocket, {:gen_tcp, #Port<0.95>, :tls_connection, :undefined}, [#PID<0.606.0>, #PID<0.605.0>]}, state: :open, streams: %{}, transport: Mint.Core.Transport.SSL, window_size: 2147483647}, hostname: "robynthinks.wordpress.com", port: 443, protocol: :https}, hostname: "robynthinks.wordpress.com", port: 443, protocol: "https", reply_tos: %{#Reference<0.2892813219.510394372.50541> => #PID<0.460.0>}, responses: %{#Reference<0.2892813219.510394372.50541> => %Mojito.Response{body: [], complete: false, headers: [], status_code: nil}}}, {:error, #Reference<0.2892813219.510394372.50541>, %Mint.HTTPError{module: Mint.HTTP2, reason: {:server_closed_request, :protocol_error}}})
    (mojito) lib/mojito/conn_server.ex:141: Mojito.ConnServer.apply_resps/2
    (mojito) lib/mojito/conn_server.ex:114: Mojito.ConnServer.handle_info/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:ssl, {:sslsocket, {:gen_tcp, #Port<0.95>, :tls_connection, :undefined}, [#PID<0.606.0>, #PID<0.605.0>]}, <<0, 0, 4, 3, 0, 0, 0, 0, 3, 0, 0, 0, 1>>}
State: %{conn: %Mojito.Conn{conn: %Mint.HTTP2{buffer: "", client_settings: %{enable_push: true, max_concurrent_streams: 100, max_frame_size: 16384}, client_settings_queue: {[], []}, decode_table: %Mint.HTTP2.HPACK.Table{entries: [], length: 0, max_table_size: 4096, size: 0}, encode_table: %Mint.HTTP2.HPACK.Table{entries: [{"If-Modified-Since", "Wed, 22 May 2019 07:43:40 GMT"}], length: 1, max_table_size: 4096, size: 78}, headers_being_processed: nil, hostname: "robynthinks.wordpress.com", next_stream_id: 5, open_client_stream_count: 1, open_server_stream_count: 0, ping_queue: {[], []}, port: 443, private: %{}, ref_to_stream_id: %{#Reference<0.2892813219.510394372.50541> => 3}, scheme: "https", server_settings: %{enable_push: true, initial_window_size: 65536, max_concurrent_streams: 128, max_frame_size: 16777215, max_header_list_size: :infinity}, socket: {:sslsocket, {:gen_tcp, #Port<0.95>, :tls_connection, :undefined}, [#PID<0.606.0>, #PID<0.605.0>]}, state: :open, streams: %{3 => %{id: 3, ref: #Reference<0.2892813219.510394372.50541>, state: :half_closed_local, window_size: 65536}}, transport: Mint.Core.Transport.SSL, window_size: 2147483647}, hostname: "robynthinks.wordpress.com", port: 443, protocol: :https}, hostname: "robynthinks.wordpress.com", port: 443, protocol: "https", reply_tos: %{#Reference<0.2892813219.510394372.50541> => #PID<0.460.0>}, responses: %{#Reference<0.2892813219.510394372.50541> => %Mojito.Response{body: [], complete: false, headers: [], status_code: nil}}}
```

---

When doing the same using curl

```sh
curl -I --http2 https://robynthinks.wordpress.com/feed/ --header "If-Modified-Since: Wed, 22 May 2019 07:43:40 GMT"
```

I get this response:

```
HTTP/2 304 
server: nginx
date: Thu, 23 May 2019 19:44:29 GMT
vary: Accept-Encoding
last-modified: Wed, 22 May 2019 07:43:40 GMT
x-nc: HIT dfw 93
x-ac: 5.ams _dfw 
strict-transport-security: max-age=15552000
```

**Comment by danhuynhdev** [comment]
The error returned is `%Mint.HTTPError{module: Mint.HTTP2, reason: {:server_closed_request, :protocol_error}}`. This error happens because Mojito is currently create connection using `Mint.HTTP.connect`, but that domain is currently using `HTTP/2`.

**Comment by danhuynhdev** [comment]
The bug has been fixed by this PR https://github.com/ericmj/mint/issues/171 on mint.

**Comment by synthomat** [comment]
Updating to a newer `mint` dependency would help here, I guess.

For the time being, a `{:mint, "~> 0.4.0", override: true}` does the trick.

**Comment by gamache** [comment]
Fixed in 0.4.0 (which requires Mint 0.4.0+).
